tests/unit/test_types.py

```diff
@@ -1234,3 +1234,154 @@ def test_datatype(tpe, simple_string, json, type_name, json_value):
+            ArrayType(IntegerType()),
+        ),
+        (
+            ArrayType,
+            {
+                "element_type": {"element_type": "integer", "type": "array"},
+                "type": "array",
+            },
+            ArrayType(ArrayType(IntegerType())),
+        ),
+        (
+            StructType,
+            {
+                "type": "struct",
+                "fields": [
+                    {
+                        "name": "NESTED",
+                        "type": {
+                            "type": "struct",
+                            "fields": [
+                                {
+                                    "name": "A",
+                                    "type": "integer",
+                                    "nullable": True,
+                                },
+                                {
+                                    "name": "B",
+                                    "type": "string",
+                                    "nullable": True,
+                                },
+                            ],
+                        },
+                        "nullable": True,
+                    }
+                ],
+            },
+            StructType(
+                [
+                    StructField(
+                        "nested",
+                        StructType(
+                            [
+                                StructField("A", IntegerType()),
+                                StructField("B", StringType()),
+                            ]
+                        ),
+                    )
+                ]
+            ),
+        ),
+        (
+            StructField,
+            {
+                "name": "AA",
+                "type": "string",
+                "nullable": True,
+            },
+            StructField("AA", StringType()),
+        ),
+        (
+            StructType,
+            {
+                "type": "struct",
+                "fields": [
+                    {"name": "A", "type": "string", "nullable": True},
+                    {"name": "B", "type": "integer", "nullable": True},
+                ],
+            },
+            StructType(
+                [StructField("a", StringType()), StructField("b", IntegerType())]
+            ),
+        ),
+        (
+            StructField,
+            {
+                "name": "AA",
+                "type": "decimal(38,0)",
+                "nullable": True,
+            },
+            StructField("AA", DecimalType()),
+        ),
+        (
+            StructField,
+            {
+                "name": "AA",
+                "type": "decimal(20,10)",
+                "nullable": True,
+            },
+            StructField("AA", DecimalType(20, 10)),
+        ),
+    ],
+)
+def test_structtype_from_json(tpe, json_dict, expected_result):
+    result = tpe.from_json(json_dict)
+    assert result == expected_result
+    assert isinstance(result, tpe)
```

Comment:
I'm wondering if we can write some round-trip tests instead of re-writing a full set.

something like 
```
datatype = MapType(IntegerType(), StringType())
json_value = datatype.jsonValue()
assert datatype == MapType.fromJson(json_value)
```

Reply:
i changed the test so that it is more concise

Reply:
thank you! looks nice.